### PR TITLE
Teams improvements

### LINKED ISSUE #428: Update user edit page to show teams and links to them.
We removed the checkboxes for this, so it would be cool to show which teams a user is in, and a link to edit that.

### PR DESCRIPTION
**Pull request by jonatanklosko**
Fixes #428, #427 and #425.

New 'Team leader' group in navigation:
image

Profile as an admin:
image

Profile as a wrc leader:
image

Profile as a normal member without rights to edit teams:
image

Example of error message when membership periods of the same user overlap each other:
image

### CODE REVIEW

File: WcaOnRails/app/controllers/application_controller.rb [hunk 1]
```diff
@@ -33,7 +33,7 @@ def doorkeeper_unauthorized_render_options(error: nil)
 
   private def redirect_unless_user(action, *args)
     unless current_user && current_user.send(action, *args)
-      flash[:danger] = "You are not allowed to #{action}"
+      flash[:danger] = "You are not allowed to #{action.to_s.sub(/^can_/, '').chomp('?').humanize.downcase}"
```
Comment (jfly) [inline WcaOnRails/app/controllers/application_controller.rb]: Clever!

File: WcaOnRails/app/controllers/teams_controller.rb [hunk 2]
```diff
@@ -1,6 +1,8 @@
 class TeamsController < ApplicationController
   before_action :authenticate_user!
-  before_action -> { redirect_unless_user(:can_edit_teams?) }
+  before_action -> { redirect_unless_user(:can_manage_teams?) }, except: [:edit, :update]
+  before_action :assign_team, only: [:edit, :update]
+  before_action -> { redirect_unless_user(:can_edit_team?, @team) }, only: [:edit, :update]
```
Comment (jfly) [inline WcaOnRails/app/controllers/teams_controller.rb]: I would prefer to have a method `team_from_params` that looks up a team from the params, rather than this `assign_team`, but I personally find `before_action` rules hard to follow and kind of scary.

Comment (jonatanklosko) [inline WcaOnRails/app/controllers/teams_controller.rb]: I saw this `*_from_params` in other controllers and I thought what would better. Actually using it in this case would produce one unnecessary query for edit and update actions, but I will change this as you prefer the mentioned way.

File: WcaOnRails/app/models/team.rb [hunk 3]
```diff
@@ -2,4 +2,18 @@ class Team < ActiveRecord::Base
   has_many :team_members, dependent: :destroy
 
   accepts_nested_attributes_for :team_members, reject_if: :all_blank, allow_destroy: true
+
+  validate :membership_periods_cannot_overlap_for_single_user
+  def membership_periods_cannot_overlap_for_single_user
+    team_members.group_by(&:user).each do |user, memberships|
+      memberships.combination(2).to_h.each do |first, second|
```
Comment (jfly) [inline WcaOnRails/app/models/team.rb]: Is it really okay to call to_h here? I fear we're going to potentially skip some problematic pairs that way.

File: WcaOnRails/app/controllers/teams_controller.rb (hunk 2 above)
Comment (jfly) [inline WcaOnRails/app/controllers/teams_controller.rb]: You're right, there is a double database query with my suggestion, but I *think* I prefer it this way.

File: WcaOnRails/app/models/team.rb (hunk 3 above)
Comment (jonatanklosko) [inline WcaOnRails/app/models/team.rb]: Good catch! I wanted to improve the readability by using block with `first` and `second` arguments instead of an array but you are absolutely right.

File: WcaOnRails/app/models/team.rb [hunk 4]
```diff
@@ -2,4 +2,18 @@ class Team < ActiveRecord::Base
...
+        first_period = first.start_date..(first.end_date || Date::Infinity.new)
+        second_period = second.start_date..(second.end_date || Date::Infinity.new)
+        if first_period.overlaps? second_period
+          message = "Membership periods overlap for user #{user.name}"
+          errors[:base] << message unless errors[:base].include?(message)
```
Comment (jfly) [inline WcaOnRails/app/models/team.rb]: Since we're already looping over users, I feel like we should `break` out as soon as we find one overlapping period for that user. I believe that will allow you to remove the `unless` here as well.

Comment (jonatanklosko) [inline WcaOnRails/app/models/team.rb]: Clever!

File: WcaOnRails/app/models/user.rb [hunk 5]
```diff
@@ -323,12 +327,16 @@ def can_admin_results?
...
-  # Team leaders should be able to edit their team.
-  # See https://github.com/cubing/worldcubeassociation.org/issues/427
-  def can_edit_teams?
+  # Returns true if the user can perform every action for teams.
+  def can_manage_teams?
     admin? || board_member? || results_team?
   end
 
+  # Returns true if the user can edit the given team.
+  def can_edit_team?(team)
```
Comment (jfly) [inline WcaOnRails/app/models/user.rb]: I'm guessing what this means, but would a clearer name for this be "can change team members" or something?

File: WcaOnRails/app/views/layouts/_navigation.html.erb [hunk 6]
```diff
@@ -136,6 +136,16 @@
                 <li><%= link_to "New post", new_post_path %></li>
               <% end %>
 
+              <% unless current_user.teams_where_is_leader.empty? %>
```
Comment (jfly) [inline WcaOnRails/app/views/layouts/_navigation.html.erb]: Nice!

File: WcaOnRails/app/models/user.rb (hunk 5 above)
Comment (jonatanklosko) [inline WcaOnRails/app/models/user.rb]: Actually that means exactly that the user can edit the team so he can change its name and description too. Really isn't that name clear?

File: WcaOnRails/app/views/users/edit.html.erb [hunk 7]
```diff
@@ -172,6 +172,12 @@
       <%= f.input :region, disabled: !editable_fields.include?(:region) %>
     <% end %>
 
+    <div class="form-group">
+      <% @user.teams.each do |team| %>
+        <%= link_to team.name, edit_team_path(team), class: "btn btn-info #{"disabled" unless current_user.can_edit_team?(team)}" %>
```
Comment (jfly) [inline WcaOnRails/app/views/users/edit.html.erb]: Can you change these to plain links? When I saw the screenshots, I assumed clicking the buttons actually had a side effect. Maybe the links could say something like "is a member of TEAM", "Is not a member of TEAM".

File: WcaOnRails/app/models/user.rb (hunk 5 above)
Comment (jfly) [inline WcaOnRails/app/models/user.rb]: Ah, I didn't realize you can edit name and description as well. This is great, then.

**Comment by jfly** [comment]
Wow three birds, one stone! Some minor comments for you, but looks great otherwise. Ping me when they're resolved, and we can merge this up!

File: WcaOnRails/app/views/users/edit.html.erb (hunk 7 above)
Comment (jonatanklosko) [inline WcaOnRails/app/views/users/edit.html.erb]: Maybe something like this:
image

Comment (jfly) [inline WcaOnRails/app/views/users/edit.html.erb]: I look that a lot more

Comment (jfly) [inline WcaOnRails/app/views/users/edit.html.erb]: *like

**Comment by jonatanklosko** [comment]
@jfly I think it's ready!

**Comment by jfly** [comment]
LGTM! Could you update the screenshots in the OP?
